### ISSUE TITLE
Can't compile cup program which has a function which returns a struct

### ISSUE THREAD

**Issue by Weltspear** [mustafaquraish/cup#9]
I tried to compile the following program:
```rust
struct A{
	a: int;
	b: int;
	c: int;
}

fn a(): A{
	let a_: A;
	a_.a = 1;
	a_.b = 2;
	a_.c = 2;
	return a_;
}


fn main(){

}

```
Compiler tells me:
```
A
compiler/codegen.cup:66:9: Unsupported type size
```

**Comment by mustafaquraish** [comment]
This is a known limitation, and there are currently no plans (by me at least) of passing around structs as values to/from functions. The reason for this is so that all function return values can fit in a single x86_64 register.

Allowing passing structs as arguments / as return values would necessitate figuring out how (and where) to properly allocate the space in the caller, and then copying the whole struct into that location. 

If this is something you're interested in working on, I'm happy to provide any assistance I can, but I am personally more focussed on other projects at the moment to work on this myself.

**Comment by Weltspear** [comment]
Okay, maybe I'll try to play around with that, but I'm not really an assembly programmer or something so idk.

**Comment by Weltspear** [comment]
Would it make sense to push struct fields to the stack instead of passing a struct on the register?

**Comment by mustafaquraish** [comment]
> Would it make sense to push struct fields to the stack instead of passing a struct on the register?

Yep. let's assume the following example:

```rust
fn get_a(): A {
    let obj: A;
    ...
    return obj;
}

fn foo() {
    let a_in_foo: A = get_a();
}
```

Here, we'd want to do something like:
1. Allocate space for `a_in_foo` in the stack frame for `foo`
2. Pass an extra (hidden) parameter to `get_a`, which represents `&a_in_foo` (the address)
3. Have `return obj` copy over the struct directly into `a_in_foo` using the address.

There are some cases to be careful of, which I'm not exactly sure what the best way to handle is, such as:

```rust
fn foo(): A {  return get_a(); }
fn bar(): A {  return foo(); }
fn baz(): A {  return bar(); }
```
One way to deal with this would be to allocate / copy an intermediate object in each function, but that seems very wasteful. It would be nice to have some way to "figure out" where the object _actually_ needs to be allocated, and then just pass around pointers.

Obviously that is more complex, and doesn't need to all be implemented at once. Just make sure that any design you pick leaves room for improvements like these to be made in the future, and that at the very least such situations are handled correctly, even if it is inefficient.

**Comment by Weltspear** [comment]
A bit offtopic question but can I share ideas/suggestions in issue tracker?

**Comment by mustafaquraish** [comment]
> A bit offtopic question but can I share ideas/suggestions in issue tracker?

Sure.

**Comment by Weltspear** [comment]
Struct copying for var to var cannot be done too?

**Comment by mustafaquraish** [comment]
> Struct copying for var to var cannot be done too?

It would be nice to support that as well. I didn't mention it because that part would be fairly straightforward, we just copy the fields from the location of the first variable to the other.

**Comment by Weltspear** [comment]
> Struct copying for var to var cannot be done too?

I've just noticed that you can do something like that:
```rust
import "std/memory.cup"

struct A{
    i: i32;
}

initialize A(b: i32){
    self.i = b;
}

fn main(){
    let a: A* = new A(5);
    let b = 5;
	
    let c: A = *a;
    
    print(c.i);
}
```
Copy by dereferencing.

**Comment by mustafaquraish** [comment]
> I've just noticed that you can do ...
> Copy by dereferencing.

I don't believe I ever added any logic in the compiler that loops over all members of a struct and copies it over. Even though this passes type checking, I suspect this won't actually work for structs that don't fit in a register. In the current state of the language this should throw an error.

**Comment by Weltspear** [comment]
> > Would it make sense to push struct fields to the stack instead of passing a struct on the register?
> 
> Yep. let's assume the following example:
> 
> ```rust
> fn get_a(): A {
>     let obj: A;
>     ...
>     return obj;
> }
> 
> fn foo() {
>     let a_in_foo: A = get_a();
> }
> ```
> 
> Here, we'd want to do something like:
> 
> 1. Allocate space for `a_in_foo` in the stack frame for `foo`
> 2. Pass an extra (hidden) parameter to `get_a`, which represents `&a_in_foo` (the address)
> 3. Have `return obj` copy over the struct directly into `a_in_foo` using the address.
> 
> There are some cases to be careful of, which I'm not exactly sure what the best way to handle is, such as:
> 
> ```rust
> fn foo(): A {  return get_a(); }
> fn bar(): A {  return foo(); }
> fn baz(): A {  return bar(); }
> ```
> 
> One way to deal with this would be to allocate / copy an intermediate object in each function, but that seems very wasteful. It would be nice to have some way to "figure out" where the object _actually_ needs to be allocated, and then just pass around pointers.
> 
> Obviously that is more complex, and doesn't need to all be implemented at once. Just make sure that any design you pick leaves room for improvements like these to be made in the future, and that at the very least such situations are handled correctly, even if it is inefficient.

Btw by allocating space in stack frame do you mean just pushing onto stack or something else?

**Comment by mustafaquraish** [comment]
> Btw by allocating space in stack frame do you mean just pushing onto stack or something else?

You should be able to use the  `add_variable_to_current_block(var)` function in `parser.cup`. you're not "pushing" anything yet, we just reserve more space in the function prologue to be able to store the struct, and the compiler just tracks the offsets.

**Comment by Weltspear** [comment]
So we allocate more space during parsing of the function?

**Comment by Weltspear** [comment]
From what I see it is already done in function parsing.

**Comment by Weltspear** [comment]
I don't think I'm able to implement this.

**Comment by mustafaquraish** [comment]
> I don't think I'm able to implement this.

No worries, thanks for trying!
